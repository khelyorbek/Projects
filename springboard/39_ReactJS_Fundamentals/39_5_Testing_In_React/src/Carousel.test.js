import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";

// smoke test
it("renders without crashing", function() {
  render(<Carousel />);
});

// snapshot test
it("matches previous snapshot", function() {
  const { asFragment } = render(<Carousel />);
  expect(asFragment()).toMatchSnapshot();
})

// unit tests
it("works when you click on the right arrow", function() {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
});

it("works when you click on the left arrow", function() {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // move forward in the carousel to 2nd image
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();

  // move left in the carousel
  const leftArrow = queryByTestId("left-arrow");
  fireEvent.click(leftArrow);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();
});

it("hides and shows arrows appropriately", function() {
  const { getByTestId } = render(<Carousel />);
  const leftArrow = getByTestId("left-arrow");
  const rightArrow = getByTestId("right-arrow");

  // expect the left arrow to be missing, but the right button to be present.
  expect(leftArrow).toHaveStyle('visibility: hidden');
  expect(rightArrow).not.toHaveStyle('visibility: hidden');

  // move forward, expect both arrow to exist
  fireEvent.click(getByTestId("right-arrow"));
 // expect the left arrow to be missing, but the right button to be present.
 expect(leftArrow).not.toHaveStyle('visibility: hidden');
 expect(rightArrow).not.toHaveStyle('visibility: hidden');

  // move forward again, expect only the right arrow to be missing
  fireEvent.click(rightArrow);
 expect(leftArrow).not.toHaveStyle('visibility: hidden');
 expect(rightArrow).toHaveStyle('visibility: hidden');

});

// PASS  src/Carousel.test.js (54.133 s)
// ✓ renders without crashing (23 ms)
// ✓ matches previous snapshot (20 ms)
// ✓ works when you click on the right arrow (20 ms)
// ✓ works when you click on the left arrow (8 ms)
// ✓ hides and shows arrows appropriately (37 ms)

// Test Suites: 1 passed, 1 total
// Tests:       5 passed, 5 total
// Snapshots:   1 passed, 1 total
// Time:        77.071 s
// Ran all test suites matching /Carousel.test.js/i