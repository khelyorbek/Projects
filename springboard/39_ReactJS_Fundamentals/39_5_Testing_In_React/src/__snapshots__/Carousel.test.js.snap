// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches previous snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="fas fa-chevron-circle-left fa-2x"
        data-testid="left-arrow"
        style="visibility: hidden;"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          Photo by Richard Pasquarella on Unsplash
        </h4>
        <img
          alt="Photo by Richard Pasquarella on Unsplash"
          class="Card-image"
          src="image1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
        style="visibility: visible;"
      />
    </div>
  </div>
</DocumentFragment>
`;
